File: myhome/occupancy/occupancy.go
package occupancy

import (
	"context"
	"encoding/json"
	"fmt"
	"myhome/mqtt"
	mqttclient "myhome/mqtt"
	"pkg/sfr"
	"strings"
	"sync/atomic"
	"time"

	"github.com/go-logr/logr"
)

// Service implements a simple HTTP server that reports home occupancy based on
// recent motion/input events observed on MQTT and mobile device presence on the LAN.
//
// Heuristic:
// - Subscribe to Shelly Gen2 events: "+/events/rpc" for NotifyStatus messages
//   containing input state changes (e.g., button presses, motion sensors)
// - Any NotifyStatus event with "input:N" updates the lastEvent timestamp
// - Poll SFR Box LAN hosts for devices whose names match configured mobile devices
// - Occupied if:
//   * now - lastEvent <= window, OR
//   * Mobile device seen online in the past window
//
// HTTP:
// - GET /status -> {"occupied":true|false,"reason":"input|mobile|mobile+input|none"}
// - Optional query: window=<seconds> to override window for this check only.

type Service struct {
	ctx              context.Context
	log              logr.Logger
	mc               mqttclient.Client
	lastEvent        atomic.Int64 // unix nano of last relevant input event
	lastMobileSeen   atomic.Int64 // unix nano of last mobile device presence
	lastSeenTime     atomic.Int64 // unix nano of last seen event (for timer expiry message)
	lastSeenTimer    *time.Timer  // timer for publishing false occupancy
	lastSeenWindow   time.Duration
	mobilePollPeriod time.Duration
	mobileDevices    []string // list of device name patterns to check for (case-insensitive substring match)
}

func NewService(ctx context.Context, log logr.Logger, mc mqttclient.Client, window time.Duration, mobilePollPeriod time.Duration, mobileDevices []string) *Service {
	s := &Service{
		ctx:              ctx,
		log:              log.WithName("occupancy.Service"),
		mc:               mc,
		lastSeenWindow:   window,
		mobilePollPeriod: mobilePollPeriod,
		mobileDevices:    mobileDevices,
	}
	return s
}

// SetWindow configures the time window for input devices & mobile device presence detection.
// If a mobile device was seen online within this window, the home is considered occupied.
func (s *Service) SetWindow(window time.Duration) *Service {
	s.lastSeenWindow = window
	return s
}

// SetMobilePollPeriod configures how often to poll the SFR Box for mobile device presence.
func (s *Service) SetMobilePollPeriod(period time.Duration) *Service {
	s.mobilePollPeriod = period
	return s
}

// SetMobileDevices configures the list of device name patterns to check for presence.
// Device names are matched case-insensitively using substring matching.
// Example: []string{"iPhone", "iPad", "Android"}
func (s *Service) SetMobileDevices(devices []string) *Service {
	s.mobileDevices = devices
	return s
}

// Start runs the MQTT subscriptions, mobile device polling
func (s *Service) Start() error {
	// Subscribe to Shelly Gen2 input status topics
	go s.subscribeInputs()

	// Start mobile device presence polling
	go s.pollMobilePresence()

	go func() {
		<-s.ctx.Done()
		if s.lastSeenTimer != nil {
			s.lastSeenTimer.Stop()
		}
	}()

	s.log.Info("Occupancy service started", "lastSeenWindow", s.lastSeenWindow, "mobilePollPeriod", s.mobilePollPeriod, "mobileDevices", s.mobileDevices)
	return nil
}

type occupancy struct {
	Occupied bool   `json:"occupied"`
	Reason   string `json:"reason,omitempty"`
}

func (s *Service) occupied(reason string) {
	// Stop existing timer if any
	if s.lastSeenTimer != nil {
		s.lastSeenTimer.Stop()
		s.lastSeenTimer = nil
	}

	status := occupancy{
		Occupied: true,
		Reason:   reason,
	}
	b, err := json.Marshal(status)
	if err != nil {
		s.log.Error(err, "Failed to marshal occupancy")
		return
	}
	s.mc.Publish(s.ctx, "myhome/occupancy", b, mqtt.AtLeastOnce, true /*retain*/, "myhome/occupancy")

	s.lastSeenTime.Store(time.Now().UnixNano())
	s.lastSeenTimer = time.AfterFunc(s.lastSeenWindow, func() {
		// Timer expired - publish false occupancy with elapsed time
		lastSeen := time.Unix(0, s.lastSeenTime.Load())
		elapsed := time.Since(lastSeen)
		hours := int(elapsed.Hours())
		minutes := int(elapsed.Minutes()) % 60
		seconds := int(elapsed.Seconds()) % 60

		var elapsedStr string
		if hours > 0 {
			elapsedStr = fmt.Sprintf("%dh %dm %ds ago", hours, minutes, seconds)
		} else if minutes > 0 {
			elapsedStr = fmt.Sprintf("%dm %ds ago", minutes, seconds)
		} else {
			elapsedStr = fmt.Sprintf("%ds ago", seconds)
		}

		status := occupancy{
			Occupied: false,
			Reason:   fmt.Sprintf("last seen %s", elapsedStr),
		}
		b, err := json.Marshal(status)
		if err != nil {
			s.log.Error(err, "Failed to marshal occupancy on timer expiry")
			return
		}
		s.mc.Publish(s.ctx, "myhome/occupancy", b, mqtt.AtLeastOnce, true /*retain*/, "myhome/occupancy")
		s.log.Info("Published false occupancy after timer expiry", "elapsed", elapsedStr)
	})
}

func (s *Service) subscribeInputs() {
	// Gen2 events/rpc for NotifyStatus with input activity
	topic := "+/events/rpc"
	ch, err := s.mc.SubscribeWithTopic(s.ctx, topic, 8, "myhome/occupancy")
	if err != nil {
		s.log.Error(err, "Failed to subscribe to events", "topic", topic)
		return
	}
	s.log.Info("Subscribed to events", "topic", topic)

	// Monitor for NotifyStatus events containing input state changes
	// Example: {"method":"NotifyStatus","params":{"input:0":{"id":0,"state":true}}}
	for {
		select {
		case <-s.ctx.Done():
			return
		case msg, ok := <-ch:
			if !ok {
				return
			}
			// Filter for NotifyStatus events with input state changes
			payload := string(msg.Payload())
			// s.log.Info("Received event", "payload", payload)
			if strings.Contains(payload, "\"NotifyStatus\"") && strings.Contains(payload, "\"input:") {
				s.log.Info("Received event with input state change", "payload", payload)
				s.lastEvent.Store(time.Now().UnixNano())
				s.occupied(fmt.Sprintf("input change: %s", payload))
			}
		}
	}
}

func (s *Service) pollMobilePresence() {
	ticker := time.NewTicker(s.mobilePollPeriod)
	defer ticker.Stop()

	// Do an immediate check on startup
	s.checkMobilePresence()

	for {
		select {
		case <-s.ctx.Done():
			return
		case <-ticker.C:
			s.checkMobilePresence()
		}
	}
}

func (s *Service) checkMobilePresence() {
	hosts, err := sfr.GetHostsList()
	if err != nil {
		s.log.Error(err, "Failed to get SFR LAN hosts")
		return
	}

	// Look for any host whose name matches one of the configured mobile devices
	for _, host := range *hosts {
		hostNameLower := strings.ToLower(host.Name)
		for _, devicePattern := range s.mobileDevices {
			if strings.Contains(hostNameLower, strings.ToLower(devicePattern)) {
				// Check if device is online (status="online" or alive > 0)
				if strings.ToLower(host.Status) == "online" || host.Alive > 0 {
					s.log.Info("Mobile device detected online", "name", host.Name, "pattern", devicePattern, "ip", host.Ip, "mac", host.Mac, "status", host.Status, "alive", host.Alive)
					s.lastMobileSeen.Store(time.Now().UnixNano())
					s.occupied(fmt.Sprintf("seen mobile: %s (%s) on %s", host.Name, host.Mac, time.Now().Format("2006-01-02 15:04:05")))
					return
				}
			}
		}
	}

	s.log.V(1).Info("No mobile devices found online", "patterns", s.mobileDevices)
}

// Start launches the occupancy HTTP service listening on port with the given MQTT client.
func Start(ctx context.Context, port int, mc mqttclient.Client, window time.Duration, mobilePollPeriod time.Duration, mobileDevices []string) error {
	log := logr.FromContextOrDiscard(ctx)
	svc := NewService(ctx, log, mc, window, mobilePollPeriod, mobileDevices)
	if mobileDevices != nil {
		svc.SetMobileDevices(mobileDevices)
	}
	return svc.Start()
}

File: internal/shelly/scripts/occupancy.js
// occupancy.js — home occupancy detection workflow (daemon-hosted).
//
// JS port of the Go occupancy service (myhome/occupancy): listed in
// daemon.scripts.run it REPLACES the Go implementation (both publish the
// retained myhome/occupancy topic).
//
// Heuristic (same as Go):
// - any Gen2 NotifyStatus event with an "input:N" change marks the home
//   occupied (button press, motion detector wired to an input, ...)
// - a configured mobile device seen online on the LAN marks the home occupied
//   (LAN polling is Go infrastructure, reached via MyHome.call("lan.hosts"))
// - occupancy expires after a quiet window; expiry publishes occupied=false
//
// Publishes (retained): myhome/occupancy = {"occupied":bool,"reason":string}
// RPC: occupancy.getstatus -> {"occupied":bool} (same verb as the Go service)
//
// KVS configuration (persisted in scripts-state/occupancy.json):
//   script/occupancy/window-hours     — quiet window before un-occupied (default 12)
//   script/occupancy/poll-minutes     — LAN presence poll period (default 5)
//   script/occupancy/mobile-patterns  — comma-separated device name patterns (default "iPhone")

var CONFIG = {
  windowMs: 12 * 3600 * 1000,
  pollMs: 5 * 60 * 1000,
  patterns: ["iPhone"],
  topic: "myhome/occupancy"
};

var STATE = {
  lastSeen: 0,
  occupied: false,
  expiryTimer: null
};

function publishStatus(occupied, reason) {
  STATE.occupied = occupied;
  var msg = JSON.stringify({ occupied: occupied, reason: reason });
  MQTT.publish(CONFIG.topic, msg, 1, true);
  MyHome.log("occupancy: " + msg);
}

function formatElapsed(ms) {
  var s = Math.floor(ms / 1000);
  var h = Math.floor(s / 3600);
  var m = Math.floor(s / 60) % 60;
  var sec = s % 60;
  if (h > 0) return h + "h " + m + "m " + sec + "s ago";
  if (m > 0) return m + "m " + sec + "s ago";
  return sec + "s ago";
}

function onWindowExpired() {
  STATE.expiryTimer = null;
  publishStatus(false, "last seen " + formatElapsed(Date.now() - STATE.lastSeen));
}

function markOccupied(reason) {
  STATE.lastSeen = Date.now();
  if (STATE.expiryTimer !== null) {
    Timer.clear(STATE.expiryTimer);
    STATE.expiryTimer = null;
  }
  publishStatus(true, reason);
  STATE.expiryTimer = Timer.set(CONFIG.windowMs, false, onWindowExpired);
}

// Gen2 events: {"method":"NotifyStatus","params":{"input:0":{...}}}
function onDeviceEvent(topic, message) {
  var payload = String(message);
  if (payload.indexOf('"NotifyStatus"') !== -1 && payload.indexOf('"input:') !== -1) {
    markOccupied("input change: " + payload);
  }
}

function matchesPattern(name) {
  var lower = String(name).toLowerCase();
  for (var i = 0; i < CONFIG.patterns.length; i++) {
    if (lower.indexOf(CONFIG.patterns[i].toLowerCase()) !== -1) return true;
  }
  return false;
}

function onLanHosts(result, error_code, error_message) {
  if (error_code !== 0) {
    MyHome.log("lan.hosts failed: " + error_message);
    return;
  }
  if (!result || !("hosts" in result) || !result.hosts) return;
  for (var i = 0; i < result.hosts.length; i++) {
    var h = result.hosts[i];
    if (!matchesPattern(h.name)) continue;
    if (String(h.status).toLowerCase() === "online" || h.alive > 0) {
      markOccupied("seen mobile: " + h.name + " (" + h.mac + ")");
      return;
    }
  }
}

function pollMobilePresence() {
  MyHome.call("lan.hosts", null, onLanHosts);
}

function onGetStatus(params) {
  return { occupied: STATE.occupied };
}

// ---- KVS-backed configuration (3 sequential loads, then start) ----

function applyConfigValue(key, value) {
  var v = String(value);
  if (key === "window-hours" && Number(v) > 0) {
    CONFIG.windowMs = Number(v) * 3600 * 1000;
  } else if (key === "poll-minutes" && Number(v) > 0) {
    CONFIG.pollMs = Number(v) * 60 * 1000;
  } else if (key === "mobile-patterns" && v.length > 0) {
    var parts = v.split(",");
    var patterns = [];
    for (var i = 0; i < parts.length; i++) {
      var p = parts[i].replace(/^\s+|\s+$/g, "");
      if (p.length > 0) patterns.push(p);
    }
    if (patterns.length > 0) CONFIG.patterns = patterns;
  }
}

var CONFIG_KEYS = ["window-hours", "poll-minutes", "mobile-patterns"];
var configIndex = 0;

function onConfigLoaded(result, error_code, error_message) {
  if (error_code === 0 && result && ("value" in result) && result.value !== null) {
    applyConfigValue(CONFIG_KEYS[configIndex], result.value);
  }
  configIndex++;
  loadNextConfig();
}

function loadNextConfig() {
  if (configIndex >= CONFIG_KEYS.length) {
    start();
    return;
  }
  Shelly.call("KVS.Get", { key: "script/occupancy/" + CONFIG_KEYS[configIndex] }, onConfigLoaded);
}

function start() {
  MQTT.subscribe("+/events/rpc", onDeviceEvent);
  Timer.set(CONFIG.pollMs, true, pollMobilePresence);
  Timer.set(3000, false, pollMobilePresence);
  MyHome.registerVerb("occupancy.getstatus", onGetStatus);
  MyHome.log("occupancy workflow started: window=" + (CONFIG.windowMs / 3600000) +
    "h poll=" + (CONFIG.pollMs / 60000) + "m patterns=" + CONFIG.patterns.join(","));
}

function main() {
  if (typeof MyHome === "undefined") {
    print("occupancy.js is a daemon workflow; it does nothing on a device");
    return;
  }
  loadNextConfig();
}

main();
